scripts/publish.py

```diff
@@ -0,0 +1,123 @@
+
+# TODO generate this by parsing toml files
+dep_graph = {
+    "wasmer-types": set([]),
+    "wasmer-vm": set(["wasmer-types"]),
+    "wasmer-compiler": set(["wasmer-vm", "wasmer-types"]),
+    "wasmer-object": set(["wasmer-types", "wasmer-compiler"]),
+    "wasmer-engine": set(["wasmer-types", "wasmer-vm", "wasmer-compiler"]),
+    "wasmer-compiler-singlepass": set(["wasmer-types", "wasmer-vm", "wasmer-compiler"]),
+    "wasmer-compiler-cranelift": set(["wasmer-types", "wasmer-vm", "wasmer-compiler"]),
+    "wasmer-compiler-llvm": set(["wasmer-types", "wasmer-vm", "wasmer-compiler"]),
+    "wasmer-engine-jit": set(["wasmer-types", "wasmer-vm", "wasmer-compiler", "wasmer-engine"]),
+    "wasmer-engine-native": set(["wasmer-types", "wasmer-vm", "wasmer-compiler", "wasmer-engine",
+                                 "wasmer-object"]),
+    "wasmer": set(["wasmer-vm", "wasmer-compiler-singlepass", "wasmer-compiler-cranelift",
+                   "wasmer-compiler-llvm", "wasmer-compiler", "wasmer-engine", "wasmer-engine-jit",
+                   "wasmer-engine-native", "wasmer-types"]),
+    "wasmer-cache": set(["wasmer"]),
+    "wasmer-wasi": set(["wasmer"]),
+    "wasmer-wasi-experimental-io-devices": set(["wasmer-wasi"]),
+    "wasmer-emscripten": set(["wasmer"]),
+}
+
+# where each crate is located in `lib`
+# this can also be generated from the toml files
+location = {
+    "wasmer-types": "wasmer-types",
+    "wasmer-vm": "vm",
+    "wasmer-compiler": "compiler",
+    "wasmer-object": "object",
+    "wasmer-engine": "engine",
+    "wasmer-compiler-singlepass": "compiler-singlepass",
+    "wasmer-compiler-cranelift": "compiler-cranelift",
+    "wasmer-compiler-llvm": "compiler-llvm",
+    "wasmer-engine": "engine",
+    "wasmer-engine-jit": "engine-jit",
+    "wasmer-engine-native": "engine-native",
+    "wasmer-cache": "cache",
+    "wasmer": "api",
+    "wasmer-wasi": "wasi",
+    "wasmer-emscripten": "emscripten",
+    "wasmer-wasi-experimental-io-devices": "wasi-experimental-io-devices",
+}
+
+dry_run = False
+
+def get_latest_version_for_crate(crate_name: str) -> Optional[str]:
+    output = subprocess.run(["cargo", "search", crate_name], capture_output=True)
+    rexp_src = '^{} = "([^"]+)"'.format(crate_name)
+    prog = re.compile(rexp_src)
+    haystack = output.stdout.decode("utf-8")
+    for line in haystack.splitlines():
+        result = prog.match(line)
+        if result:
+            return result.group(1)
+
+def is_crate_already_published(crate_name: str) -> bool:
+    found_string = get_latest_version_for_crate(crate_name)
+    if found_string is None:
+        return False
+
+    return target_version == found_string
+    
+# TODO: add dry run mode
+def publish_crate(crate: str):
+    os.chdir("lib/{}".format(location[crate]))
+
+    global dry_run
+    if dry_run:
+        print("In dry-run: not publishing crate `{}`".format(crate))
+    else:
+        output = subprocess.run(["cargo", "publish"])
+
+    os.chdir("../..")
```

Comment:
Note that this may have surprising behaviour if lib/location_crate goes through a symlink. Obviously we don't expect that so it's up to you whether to fix it.

Reply:
Good call, I updated it to store the starting dir and return to it; this also solves the problem with `deprecated/runtime-*` which aren't part of the script yet